codalab/worker/tar_file_stream.py

```diff
@@ -0,0 +1,75 @@
+from io import SEEK_SET, SEEK_CUR, SEEK_END, BytesIO
+
+from codalab.worker.un_gzip_stream import BytesBuffer
+from ratarmount import FileInfo, SQLiteIndexedTar
+
+
+class TarFileStream(BytesIO):
+    """Streams a file from a tar archive stored on Blob Storage.
+
+    The general idea is that whenever .read() is called on this class,
+    it will read the specified number of bytes through ratarmount's tf.open()
+    API on the associated file and return those bytes.
+
+    TODO (Ashwin): If we can add tf.open() support upstream to the ratarmount API
+    (right now it only supports tf.read()), we may not have a need for this class anymore.
+    """
+
+    BUFFER_SIZE = 100 * 1024 * 1024  # Read in chunks of 100MB
```

Comment:
how did you arrive at this value?

Reply:
I just used the same buffer size in the other stream class we had, `GzipStream`, which @nelson-liu initially added in https://github.com/codalab/codalab-worksheets/commit/df9a7fbb783f3ae4c5b63f4fff94e33e06a308c3. The tradeoffs here are:
- if BUFFER_SIZE is too small, the Python code may add too much overhead and slow down streaming a file if we're trying to stream a large amount at a time.
- if BUFFER_SIZE is too large, then if we just try to stream a small amount (like a couple of bytes), we'll add more overhead when trying to stream just a few bytes at a time.

Now that you mention it, though, maybe a smaller value would be better. I'll do some testing to see what the optimal value would be.

Reply:
I talked with @nelson-liu about the design of the buffer size and we came to the conclusion that it's not really necessary; it adds too much overhead by reading too much from the file if we only want to read a small number of bytes. The concern I mentioned about BUFFER_SIZE being too small doesn't seem to be a problem because when a file is streamed from a REST endpoint, it's read 8192 bytes at a time.